plutus-metatheory/Setup.hs

```diff
@@ -1,49 +1,109 @@
-That means we need a pre-build hook. But then it won't run before Haddock!
+Finally, the order in which the modules are listed in exposed-modules matters a lot!
+The MAlonzo.Code files must be listed first, otherwise cabal will fail with:
+setup: can't find source for MAlonzo/Code/* in src
 
-So: we do both! Since Agda is reasonably good at not redoing work, we mostly don't
-pay the cost twice on the first configure/build sequence.
+TODO Newer (> 3.6.3.0) versions of the cabal library introduced a ppOrdering field in PreProcessor
+which can reorder all modules.
+https://hackage.haskell.org/package/Cabal-3.8.1.0/docs/Distribution-Simple-PreProcess.html
+Until we upgrade cabal, we just have to be careful to expose our modules in the right order.
+Once cabal is upgrdaded, we can implement ppOrdering as reorderModules:
+
+reorderModules :: Verbosity -> [FilePath] -> [ModuleName] -> IO [ModuleName]
+reorderModules _ _ = sortBy malonzoCodeOrdering
+  where
+    malonzoCodeOrdering :: ModuleName -> ModuleName -> Ordering
+    malonzoCodeOrdering name _
+      | "MAlonzo.Code" `isPrefixOf` name = LT
+      | otherwise = EQ
 -}
 
-postConf' :: Args -> ConfigFlags -> PackageDescription -> LocalBuildInfo -> IO ()
-postConf' _ _ _ _ = runAgda
+data AgdaProgramStatus
+  = Run
+  | NotRun
+  deriving (Eq, Ord, Show)
+
+
+agdaProgramStatus :: IORef AgdaProgramStatus
+agdaProgramStatus = unsafePerformIO (newIORef NotRun)
+{-# NOINLINE agdaProgramStatus #-}
+
+
+main :: IO ()
+main = D.defaultMainWithHooks userHooks
+  where
+    userHooks :: D.UserHooks
+    userHooks = D.simpleUserHooks { D.hookedPreProcessors = preProcessors }
+
+    preProcessors :: [D.PPSuffixHandler]
+    preProcessors = [("lagda", agdaPreProcessor)]
+
+
+agdaPreProcessor :: D.BuildInfo -> D.LocalBuildInfo -> D.ComponentLocalBuildInfo -> D.PreProcessor
+agdaPreProcessor _ lbi _ = D.PreProcessor
+  { D.platformIndependent = True
+  , D.runPreProcessor = D.mkSimplePreProcessor preProcessors
+  }
+  where
+    preProcessors :: FilePath -> FilePath -> D.Verbosity -> IO ()
+    preProcessors _ _ _ = do
+      status <- readIORef agdaProgramStatus
+      case status of
+        NotRun -> do
+          putStrLn "***** running agda preprocessor..."
+          runAgda
+          writeIORef agdaProgramStatus Run
+        Run -> do
+          putStrLn "***** agda already run, skipping preprocessor hook"
 
--- This adds a pre-build hook which calls Agda. No attempt is made to fail gracefully.
--- Doing this before building ensures that the generated Haskell files are always up-to-date, since cabal won't
--- track the dependency on the source Agda files, but Agda should do
-preBuild' :: Args -> BuildFlags -> IO ()
-preBuild' _ _ = runAgda
+    runAgda :: IO ()
+    runAgda = D.runProgram D.verbose agdaProgram
+      [ "--compile-dir", D.buildDir lbi
+      , "--compile"
+      , "--ghc-dont-call-ghc"
+      , "--local-interfaces"
```

Comment:
At the moment, this flag means that we put the generated Agda interface files next to the source files. Really, those are also auto-generated and would ideally go in `dist-newstyle` also. I don't know if there's a way to stop it doing that. I tried removing this option and it still put them next to the source files :thinking:

Reply:
Yes `--local-interfaces` seems to be a noop for me also. 🤔

Reply:
Seems like it's something to do with some agda-specific build directory: https://github.com/agda/agda/pull/3987/files#diff-06572a96a58dc510037d5efa622f9bec8519bc1beab13c9f251e97e657a9d4edR9